src/libstd/panicking.rs

```diff
@@ -372,8 +370,7 @@ pub fn begin_panic_handler(info: &PanicInfo<'_>) -> ! {
 
     let loc = info.location().unwrap(); // The current implementation always returns Some
     let msg = info.message().unwrap(); // The current implementation always returns Some
-    let file_line_col = (loc.file(), loc.line(), loc.column());
-    rust_panic_with_hook(&mut PanicPayload::new(msg), info.message(), &file_line_col);
+    rust_panic_with_hook(&mut PanicPayload::new(msg), info.message(), loc);
```

Comment:
Maybe `rust_panic_with_hook` should take a whole `PanicInfo`, instead of creating one inside?

Reply:
Looks like to avoid re-creating the `PanicInfo` we'd need to change the signature of `begin_panic_handler` which is public and stable, I think.

Here's the diff I'm trying:

```
diff --git a/src/libstd/panicking.rs b/src/libstd/panicking.rs
index e3ce7a33a6f..f28b0a2db1f 100644
--- a/src/libstd/panicking.rs
+++ b/src/libstd/panicking.rs
@@ -321,14 +321,14 @@ pub fn begin_panic_fmt(msg: &fmt::Arguments<'_>) -> ! {
         unsafe { intrinsics::abort() }
     }

-    let info = PanicInfo::internal_constructor(Some(msg), Location::caller());
-    begin_panic_handler(&info)
+    let mut info = PanicInfo::internal_constructor(Some(msg), Location::caller());
+    begin_panic_handler(&mut info)
 }

 /// Entry point of panics from the libcore crate (`panic_impl` lang item).
 #[cfg_attr(not(test), panic_handler)]
 #[unwind(allowed)]
-pub fn begin_panic_handler(info: &PanicInfo<'_>) -> ! {
+pub fn begin_panic_handler(info: &mut PanicInfo<'_>) -> ! {
     struct PanicPayload<'a> {
         inner: &'a fmt::Arguments<'a>,
         string: Option<String>,
@@ -368,9 +368,7 @@ pub fn begin_panic_handler(info: &PanicInfo<'_>) -> ! {
     // panic + OOM properly anyway (see comment in begin_panic
     // below).

-    let loc = info.location().unwrap(); // The current implementation always returns Some
-    let msg = info.message().unwrap(); // The current implementation always returns Some
-    rust_panic_with_hook(&mut PanicPayload::new(msg), info.message(), loc);
+    rust_panic_with_hook(&mut PanicPayload::new(info.message().unwrap()), info);
 }

 /// This is the entry point of panicking for the non-format-string variants of
@@ -395,7 +393,8 @@ pub fn begin_panic<M: Any + Send>(msg: M, #[cfg(bootstrap)] _: &(&str, u32, u32)
     // we do start doing this, then we should propagate this allocation to
     // be performed in the parent of this thread instead of the thread that's
     // panicking.
-    rust_panic_with_hook(&mut PanicPayload::new(msg), None, Location::caller());
+    let mut info = PanicInfo::internal_constructor(None, Location::caller());
+    rust_panic_with_hook(&mut PanicPayload::new(msg), &mut info);

     struct PanicPayload<A> {
         inner: Option<A>,
@@ -432,8 +431,7 @@ pub fn begin_panic<M: Any + Send>(msg: M, #[cfg(bootstrap)] _: &(&str, u32, u32)
 /// abort or unwind.
 fn rust_panic_with_hook(
     payload: &mut dyn BoxMeUp,
-    message: Option<&fmt::Arguments<'_>>,
-    location: &Location<'_>,
+    info: &mut PanicInfo<'_>,
 ) -> ! {
     let panics = update_panic_count(1);

@@ -451,7 +449,6 @@ fn rust_panic_with_hook(
     }

     unsafe {
-        let mut info = PanicInfo::internal_constructor(message, location);
         HOOK_LOCK.read();
         match HOOK {
             // Some platforms (like wasm) know that printing to stderr won't ever actually
```

and here's the error:

```
error: argument should be `&PanicInfo`
   --> src/libstd/panicking.rs:331:34
    |
331 | pub fn begin_panic_handler(info: &mut PanicInfo<'_>) -> ! {
    |                                  ^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

Reply:
`begin_panic_handler` is not stable, but it is marked as `#[panic_handler]` and that has requirements on the function, fixing its signature.

Also see https://github.com/rust-lang/rust/issues/66745#issuecomment-558558882 where I mused about making this a mutable reference.

Reply:
It looks like this will be another bootstrap-crossing migration. OK with y'all to do it as a follow-up with an issue filed?

Reply:
I put in an hour or so to try this out, and I'm not sure it's actually possible to pass a single mutable reference down. I added a comment to what I think is the most confusing part of the code. Let me know if I'm missing something or if it's unclear!

Reply:
You can't just change this! `#[panic_handler]` is stable and any change to its rules requires an RFC.